cmd/clusterctl/client/repository/repository_gitlab.go

```diff
@@ -0,0 +1,190 @@
+func NewGitLabRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient, opts ...gitLabRepositoryOption) (Repository, error) {
+	if configVariablesClient == nil {
+		return nil, errors.New("invalid arguments: configVariablesClient can't be nil")
+	}
+
+	rURL, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+
+	urlSplit := strings.Split(strings.TrimPrefix(rURL.RawPath, "/"), "/")
+
+	// Check if the url is a github repository
+	if rURL.Scheme != httpsScheme ||
+		len(urlSplit) != 9 ||
+		!strings.HasPrefix(rURL.RawPath, gitlabPackagesAPIPrefix) ||
+		urlSplit[4] != gitlabPackagesAPIPackages ||
+		urlSplit[5] != gitlabPackagesAPIGeneric {
+		return nil, errors.New("invalid url: a GitLab repository url should be in the form https://{host}/api/v4/projects/{projectSlug}/packages/generic/{packageName}/{defaultVersion}/{componentsPath}")
+	}
+
+	httpClient := http.DefaultClient
+	// Extract all the info from url split.
+	host := rURL.Host
+	projectSlug := urlSplit[3]
+	packageName := urlSplit[6]
+	defaultVersion := urlSplit[7]
+	rootPath := "."
+	componentsPath := urlSplit[8]
+
+	repo := &gitLabRepository{
+		providerConfig:        providerConfig,
+		configVariablesClient: configVariablesClient,
+		httpClient:            httpClient,
+		host:                  host,
+		projectSlug:           projectSlug,
+		packageName:           packageName,
+		defaultVersion:        defaultVersion,
+		rootPath:              rootPath,
+		componentsPath:        componentsPath,
+	}
+
+	// process gitLabRepositoryOptions
+	for _, o := range opts {
+		o(repo)
+	}
+
+	return repo, nil
+}
+
+// Host returns host field of gitLabRepository struct.
+func (g *gitLabRepository) Host() string {
+	return g.host
+}
+
+// ProjectSlug returns projectSlug field of gitLabRepository struct.
+func (g *gitLabRepository) ProjectSlug() string {
+	return g.projectSlug
+}
+
+// PackageName returns packageName field of gitLabRepository struct.
+func (g *gitLabRepository) PackageName() string {
+	return g.packageName
+}
+
+// DefaultVersion returns defaultVersion field of gitLabRepository struct.
+func (g *gitLabRepository) DefaultVersion() string {
+	return g.defaultVersion
+}
+
+// GetVersions returns the list of versions that are available in a provider repository.
+func (g *gitLabRepository) GetVersions() ([]string, error) {
+	// FIXME Get versions from GitLab API
+	return []string{g.defaultVersion}, nil
+}
+
+// RootPath returns rootPath field of gitLabRepository struct.
+func (g *gitLabRepository) RootPath() string {
+	return g.rootPath
+}
+
+// ComponentsPath returns componentsPath field of gitLabRepository struct.
+func (g *gitLabRepository) ComponentsPath() string {
+	return g.componentsPath
+}
+
+// GetFile returns a file for a given provider version.
+func (g *gitLabRepository) GetFile(version, path string) ([]byte, error) {
+	ctx := context.TODO()
```

Comment:
Q: What about using a context with timeout, so the http request does not wait indefinetly